src/libcore/char/methods.rs

```diff
@@ -1673,3 +1635,60 @@ impl char {
         }
     }
 }
+
+#[inline]
+fn len_utf8(code: u32) -> usize {
+    if code < MAX_ONE_B {
+        1
+    } else if code < MAX_TWO_B {
+        2
+    } else if code < MAX_THREE_B {
+        3
+    } else {
+        4
+    }
+}
+
+/// Encodes a raw u32 value as UTF-8 into the provided byte buffer,
+/// and then returns the subslice of the buffer that contains the encoded character.
+///
+/// Unlike `char::encode_utf8`, this method can be called on codepoints in the surrogate range.
```

Comment:
What does it do on those codepoints -- does it produce valid UTF-8? Based on the UB issue it sounds like yes. Can you add that to this comment? Maybe a brief "note that `char` has a validity invariant that excludes surrogate values" too.

Reply:
Is it valid UTF-8 when a surrogate pair gets encoded? I don't know.

Reply:
Good point.

This function codes a code point in [generalized UTF-8](https://simonsapin.github.io/wtf-8/#generalized-utf8). For surrogate code points, the resulting byte sequence is not well-formed UTF-8 and therefore not valid `str`. The return type should be changed to `&mut [u8]` and the `from_utf8_unchecked` call moved to `char::encode_utf8`.

Reply:
> The return type should be changed to &mut [u8] and the from_utf8_unchecked call moved to char::encode_utf8.

Done.